sql/analyzer/resolve_ctes.go

```diff
@@ -280,104 +283,100 @@ func hoistRecursiveCte(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope,
+// "the recursive table must be referenced only once, and not in any subquery"
+func splitRecursiveCteUnion(name string, n sql.Node) (sql.Node, sql.Node) {
+	union, ok := n.(*plan.Union)
+	if !ok {
+		return n, nil
 	}
-	switch n := union.Right().(type) {
-	case *plan.Project, *plan.GroupBy:
+	switch union.Right().(type) {
+	case *plan.Union:
+		panic("not supported")
+	case *plan.SubqueryAlias:
+		// can't be recursive
+		return plan.NewUnion(union.Left(), union.Right(), union.Distinct, union.Limit, union.SortFields), nil
 	default:
-		return nil, sql.ErrInvalidRecursiveCteRecursiveQuery.New(n)
 	}
 
-	return plan.NewRecursiveCte(union.Left(), union.Right(), sq.Name(), sq.Columns, deduplicate), nil
-}
-
-func resolveRecursiveCte(ctx *sql.Context, a *Analyzer, node sql.Node, sq sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
-	rCte := node.(*plan.RecursiveCte)
-	if rCte == nil {
-		return node, transform.SameTree, nil
+	if hasTable(name, union.Right()) {
+		l, r := splitRecursiveCteUnion(name, union.Left())
+		if r == nil {
+			return union.Left(), union.Right()
+		}
+		return l, plan.NewUnion(r, union.Right(), union.Distinct, union.Limit, union.SortFields)
 	}
 
-	newInit, _, err := a.analyzeThroughBatch(ctx, rCte.Init, scope, "default-rules", sel)
-	if err != nil {
-		return node, transform.SameTree, err
-	}
+	return plan.NewUnion(union.Left(), union.Right(), union.Distinct, union.Limit, union.SortFields), nil
+}
 
-	// create recursive schema from initial projection cols and names
-	var outputProj []sql.Expression
-	switch n := newInit.(type) {
-	case *plan.Project:
-		outputProj = n.Projections
-	case *plan.GroupBy:
-		outputProj = n.SelectedExprs
+// resolveRecursiveCte resolves the static left node of the CTE to perform
+// 1) schema discovery for the CTE and recursive right half, and 2) replace
+// recursive UnresolvedTable references with resolved RecursiveTable nodes.
+func resolveRecursiveCte(
+	ctx *sql.Context,
+	a *Analyzer,
+	rCte *plan.RecursiveCte,
+	scope *Scope,
+	sel RuleSelector,
+) (sql.Node, error) {
+	newInit, _, err := a.analyzeThroughBatch(ctx, rCte.Left(), scope, "default-rules", sel)
+	if err != nil {
+		return rCte, err
 	}
 
-	schema := make(sql.Schema, len(outputProj))
-	var name string
-	for i, p := range outputProj {
-		switch c := p.(type) {
-		case *expression.Alias, *expression.GetField:
-			name = c.(sql.Nameable).Name()
-		case *expression.Literal, sql.Aggregation:
-			name = c.String()
-		default:
-			return nil, transform.SameTree, fmt.Errorf("failed to resolve or unsupported field: %v", p)
-		}
-		if i < len(rCte.Columns) {
-			name = rCte.Columns[i]
-		}
-		schema[i] = &sql.Column{
-			Name:     name,
-			Source:   rCte.Name(),
-			Type:     p.Type(),
-			Nullable: p.IsNullable(),
+	recSch := make(sql.Schema, len(newInit.Schema()))
+	for i, c := range newInit.Schema() {
+		newC := c.Copy()
+		if len(rCte.Columns) > 0 {
+			newC.Name = rCte.Columns[i]
```

Comment:
This can panic if the schema lengths aren't the same. Do we do that check elsewhere?